boards: stm32h7s78_dk: Introduce App in Ext Flash variant

erwango opened the pull request:
Based on https://github.com/zephyrproject-rtos/zephyr/pull/97037

Similarly to what is done for `stm32h750_dk` board in #97037, introduce a `ext_flash_app` board variant aiming at being chainloaded by MCUBoot.
Requires mcuboot counter part (https://github.com/mcu-tools/mcuboot/pull/2510) to allow MCUBoot to run on internal flash.

Doc not provided, waiting for https://github.com/zephyrproject-rtos/zephyr/pull/97037 to be merged.
@JarmouniA, in a fisrst step, I plan to redirect doc to `stm32h750_dk` to avoid redundant doc in each board.

erwango committed:
boards: st: stm32h7s78_dk: Add ext_flash_app variant

Add ext_flash_app variant to stm32h7s78_dk to allow building
application running on ext flash.

Largely based on A.Jarmouni's work to enable the same on stm32h750_dk
in PR 97037.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
samples: sysbuild: with_mcuboot: stm32h7s78_dk: Use ext_flash_app variant

Now that ext_flash_app variant is available, it should be used for
mcuboot related apps on this board.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
boards: st: stm32h7s78_dk: Add debug using stlink gdbserver

Add debug using stlink gdbserver as an alternative to pyocd.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

JarmouniA commented on boards/st/stm32h7s78_dk/stm32h7s78_dk-common.dtsi:
```diff
@@ -0,0 +1,349 @@
+			compatible = "soc-nv-flash";
+			reg = <0x0 DT_SIZE_M(128)>;
+			write-block-size = <1>;
+			erase-block-size = <DT_SIZE_K(128)>;
```
```suggestion
			erase-block-size = <DT_SIZE_K(4)>;
```

JarmouniA reviewed (commented):
There are other MCUboot tests the board can be added to (see the H750B-DK PR).

JarmouniA commented on boards/st/stm32h7s78_dk/stm32h7s78_dk_stm32h7s7xx_ext_flash_app.dts:
```diff
@@ -0,0 +1,55 @@
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		boot_partition: partition@0 {
```
this is already in the common dtsi.

erwango commented on boards/st/stm32h7s78_dk/stm32h7s78_dk-common.dtsi:
```diff
@@ -0,0 +1,349 @@
+		reg = <0x70000000 DT_SIZE_M(64)>;
+		zephyr,memory-region = "EXTMEM";
+		/* The ATTR_MPU_EXTMEM attribut causing a MPU FAULT */
+		zephyr,memory-attr = <DT_MEM_ARM(ATTR_MPU_IO)>;
```
Display related features are not working yet when executed from external flash. Probably due to `STM32_LTDC_FB_USE_SHARED_MULTI_HEAP`, had no time to investigate this (likely related to use of multiple XSPI instances used in //)

Is everyone fine to put this as a notable issue that can still be investigated and fixed in RC ?

JarmouniA replied:
> Display related features are not working yet

Weird, even `samples/drivers/display`?

Non-blocking, but could you try with `MPU_FLASH` attribute?

etienne-lms commented on boards/st/stm32h7s78_dk/stm32h7s78_dk_stm32h7s7xx_ext_flash_app.yaml:
```diff
@@ -0,0 +1,20 @@
+arch: arm
+toolchain:
+  - zephyr
+ram: 640
```
`ram: 128` as per `sram0` current size?
(Maybe a appended commit to fix *boards/st/stm32h7s78_dk/stm32h7s78_dk.yaml* the same way)

erwango replied:
yes, good catch.

erwango committed:
boards: st: stm32h7s78_dk: Fix ram size in board yaml file

Actual available internal RAM size was wrong, fix it.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

nordicjm commented on boards/st/stm32h7s78_dk/Kconfig.sysbuild:
```diff
@@ -0,0 +1,18 @@
+endchoice
+
+choice MCUBOOT_MODE
+	default MCUBOOT_MODE_SINGLE_APP
```
this can use direct-xip if wanted, there is no limitation on mode that can be used

erwango commented on west.yml:
```diff
@@ -23,6 +23,8 @@ manifest:
       url-base: https://github.com/zephyrproject-rtos
     - name: babblesim
       url-base: https://github.com/BabbleSim
+    - name: mcuboot_upstream
```
@nordicjm Don't you need to approve & merge https://github.com/mcu-tools/mcuboot/pull/2510 first ?

nordicjm replied:
seems it does eventually work, use the commit from here: https://github.com/nordicjm/zephyr/tree/somemcubootupdatecommits

nordicjm committed:
west.yml: MCUboot synchronization from upstream

Update Zephyr fork of MCUboot to revision:
  96576b341ee19f1c3af6622256b0d4f3d408e1e3

Brings following Zephyr relevant fixes:

  - 96576b34 boot: zephyr: socs: stm32h7s7xx: Add support for
    ext_flash_app variant
  - 56538cf6 bootutil: Move update-independent code
  - fccd8905 loader: Move boot_get_max_app_size(..) API
  - 0248a023 bootutil: Move state-independent area APIs
  - f2d3f00a loader: Unify image check API.
  - dcc66e51 loader: Unify header_valid(..) API
  - 257265c8 loader: Optimize boot_check_header_erased(..)
  - 76e56e4f loader: Rename boot_version_cmp
  - 5311589b loader: Fix compile-time issues in loader.c
  - 606a1934 boot: zephyr: socs: add overlay & conf for
    stm32h573xx_ext_flash_app
  - 521fc0bf bootutil: Conditionally include mbedtls/ oid.h,
    asn1.h
  - 71b41e38 boot: zephyr: boards: remove nrf54h20dk overlay
  - 2fc1bd84 bootutil: Drop slot number and boot_state from most
    boot_enc functions
  - 1dd8ae60 boot: zephyr: rework stm32h750b_dk board overlay &
    conf

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

etienne-lms reviewed (approved):
LGTM
